Add all-electric plant model to DHC package

AntoineGautier opened the pull request:
This closes #3188.

Two models fail to compile with OMC, see `.conf.yml`:
```yml
- model_name: Buildings.Experimental.DHC.Plants.Combined.Controls.BaseClasses.Validation.StagingPlant
  openmodelica:
    comment: Internal error BackendDAETransform.analyseStrongComponentBlock failed (Purely discrete algebraic loops cannot be solved by iterative processes. Try to break them open using the delay() operator.)
    translate: false
- model_name: Buildings.Experimental.DHC.Plants.Combined.Validation.AllElectricCWStorage
  openmodelica:
    comment: "Internal error NFCeval.evalBinaryDiv failed to evaluate \u20181.0 / pla.chiHea.COPCasCoo_nominal\u2018"
    translate: false
```

:warning: This PR also introduces some changes in the graphical annotations of the following interface classes: `PartialETS`, `PartialBuilding`, `PartialBuildingWithPartialETS`, `PartialPlant`. The color convention for fluid streams previously used was confusing: for instance CHW return was represented in red color and CHW supply in blue color. The change consists in representing both the CHW return and CHW supply in blue color. In addition, the diagram layer for `PartialETS` is expanded as the current model showed a need for more real estate.

AntoineGautier commented:
Todo before merging:

- [ ] Add sizing considerations to documentation
- [ ] Add icons to components
- [ ] Increase the size of the diagram layer

dhblum commented:
@AntoineGautier I was thinking about a validation model that has both heating and cooling loads from buildings in a district model.  There's an example cooling district in ``Buildings.Experimental.DHC.Examples.Cooling.ElectricChillersDirectETS``, however, right now there is no equivalent on the heating side.  In particular, we don't have anything similar to ``Buildings.Experimental.DHC.Loads.Cooling.BuildingTimeSeriesWithETS``, use direct or indirect ETS from ``Buildings.Experimental.DHC.EnergyTransferStations.Cooling``, on the heating side.  The examples we have utilize heating from ambient loops, which use ``Buildings.Experimental.DHC.EnergyTransferStations.Combined.HeatPumpHeatExchanger``, or there is currently a development branch with steam loads, but these use steam-based ETS. Put another way, in terms of ``Buildings.Experimental.DHC.Types.DistrictSystemType``, I don't think we have a load for type ``HeatingGeneration2to4`` or ``CombinedGeneration2to4``.  I think it could be worthwhile to implement as part of a new validation model for this plant.

dhblum reviewed (changes requested):
Thanks again @AntoineGautier.  This generally looks good.  I removed one section of duplicate documentation I found and also merged the latest master.  Unit tests still passed locally.  Two questions/comments:

- [ ] In ``Buildings.Experimental.DHC.Plants.Combined.Controls.BaseClasses.TankCycle``, I think it would be helpful to provide a brief description of the operating intent/reason of each of the two tank cycles.   
- [ ] In ``Buildings.Experimental.DHC.Plants.Combined.Validation.AllElectricCWStorage``, why is ``pla.nPumConWatCon = 2`` while all other pump numbers in ``pla`` are equal to the parameters indicating the number of units the pumps are serving?  Not sure if this is on purpose.

AntoineGautier commented:
@dhblum Thanks for looking into that so quickly. I added the rationale for the tank temperature cycles. Regarding the number of pumps : this is really a designer choice I would say (for headered pumps with VFDs). For instance TE use 2 pumps for their 2+3 TIER plant. In my last commit I changed the defaults to reduce the number of CHW pumps and assigned the same number of CWC pumps. The default is used in the validation model as well.

dhblum commented:
In the latest commits I've:
- added an example model of a district network with two buildings served by the new plant
- added direct and indirect ETS and BuildingTimeSeriesWithETS for 2to4Gen heating
- resolved master conflicts
- updated unit test results
- added new example model to exempt list for OpenModelica and Optimica with comments for latest versions

Still to do:
- [ ] Update check valve pressure drop after https://github.com/lbl-srg/modelica-buildings/pull/3390 merged to master, and associated reference results
- [ ] Update release notes
- [ ] Review

dhblum commented:
@mwetter Aside from the to-dos I listed, I think this is largely ready for your review.

FYI @AntoineGautier feel free to review my updates as well.

mwetter reviewed (changes requested):
Please see the inline comments.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Controls/BaseClasses/DirectHeatRecovery.mo:
```diff
@@ -0,0 +1,265 @@
+    annotation(Dialog(group="HW loop and heat recovery chillers"));
+  parameter Modelica.Units.SI.MassFlowRate mChiWatChiHea_flow_min
+    "HRC CHW minimum mass flow rate (each unit)"
+    annotation(Dialog(group="HW loop and heat recovery chillers"));
```
Resolved in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/f437afad024f5d1aea73581c3a535b4d4309ddf5.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Subsystems/BaseClasses/MultipleValves.mo:
```diff
@@ -0,0 +1,170 @@
+
+  parameter Boolean use_inputFilter=energyDynamics<>Modelica.Fluid.Types.Dynamics.SteadyState
+    "= true, if opening is filtered with a 2nd order CriticalDamping filter"
+    annotation(Dialog(tab="Dynamics", group="Filtered opening"));
```
Resolved in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/3b3c6584479bc8281f0a22cc9f96d80c45ce2c96.

dhblum commented on Buildings/Experimental/DHC/EnergyTransferStations/BaseClasses/PartialIndirect.mo:
```diff
@@ -0,0 +1,273 @@
+    annotation(Dialog(group="Nominal condition"));
+  parameter Modelica.Units.SI.MassFlowRate mBui_flow_nominal(
+    final min=0,
+    final start=0.5)
```
Addressed in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/0c90c7c543d627af573a091f5b95758bfe77ef25.

dhblum commented on Buildings/Experimental/DHC/EnergyTransferStations/BaseClasses/PartialIndirect.mo:
```diff
@@ -0,0 +1,273 @@
+          controllerType == Modelica.Blocks.Types.SimpleController.PD or
+          controllerType == Modelica.Blocks.Types.SimpleController.PID));
+  parameter Real yMax(
+    final start=1)=1
```
Addressed in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/e3f480ea48f7716a9d847b49c4891f8980803f73.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Subsystems/BaseClasses/MultipleValves.mo:
I switched the valves in ``Buildings.Experimental.DHC.Plants.Combined.Subsystems.ChillerGroup`` to use a single Two Way Linear valve with ``sum(y)/nUni`` as the input:

<img width="355" alt="Screen Shot 2023-05-30 at 4 10 32 PM" src="https://github.com/lbl-srg/modelica-buildings/assets/19480999/e9a1542b-e7ef-423c-bbd8-5079849fa49f">


Simulating the model ``Buildings.Experimental.DHC.Plants.Combined.Validation.AllElectricCWStorage`` leads to similar results (thought see my comment on pump power below) with reduced cpu time:
<img width="633" alt="Screen Shot 2023-05-30 at 3 05 50 PM" src="https://github.com/lbl-srg/modelica-buildings/assets/19480999/33de2f1d-b9ef-45fd-a456-1c16a44cf4a5">

However, simulating the model ``Buildings.Experimental.DHC.Plants.Combined.Subsystems.Validation.ChillerGroup`` leads to different results for the flow through the chiller, results shown for evaporator:

<img width="672" alt="Screen Shot 2023-05-30 at 3 46 31 PM" src="https://github.com/lbl-srg/modelica-buildings/assets/19480999/9fa1c729-0311-46b7-9453-4491ef0d33fc">

Not sure why the above results differ since ``linearized=true`` for all valve models?  One thing might be the model ``Buildings.Experimental.DHC.Plants.Combined.Subsystems.Validation.ChillerGroup`` has a constant dp, whereas the implementation of the full plant has dynamic control to maintain specified flow rates or leaving temperature set points.  Turns out, in ``Buildings.Experimental.DHC.Plants.Combined.Validation.AllElectricCWStorage``, the change leads to slightly less pumping power when the valves are not both full open:

<img width="598" alt="Screen Shot 2023-05-30 at 4 08 47 PM" src="https://github.com/lbl-srg/modelica-buildings/assets/19480999/6af0d674-ad4d-46b3-b6a2-6d2f2c57c1b2">

dhblum replied:
@mwetter I tested changing the valve linearization from true to false (still keeping the original implementation of the multiple parallel valves) and simulated ``Buildings.Experimental.DHC.Plants.Combined.Validation.AllElectricCWStorage`` with the following results: Cvode failed to complete the simulation and Dassl completed but with significantly more CPU time.

<img width="597" alt="Screen Shot 2023-05-31 at 2 22 06 PM" src="https://github.com/lbl-srg/modelica-buildings/assets/19480999/70608ae6-8222-4edb-b0e0-e288b501e802">

mwetter replied:
Let's keep them linearized.

dhblum replied:
@mwetter I agree.  

Also, with respect to the assumption of flow through the parallel valves and how that corresponds to flow/loading of parallel chillers, there is indeed an assumption that all valves of operating chillers operate at the same opening position.  The chillers are "staged" on and off in coordination with the number of open valves.  Here's a revised plot of ``Buildings.Experimental.DHC.Plants.Combined.Validation.AllElectricCWStorage`` showing the coordination of the valve opening/closing with the chiller staging.  Top plot is chiller enable signal (effectively the multiplier on the flow rate through the single chiller instance), the middle plot shows the signals to the multiple valves for the evaporator, and the bottom plot shows the flow multiplier input signal for the evaporator inlet flow.  Note that the second valve is only opened when the second chiller comes online.  Note also in the middle of the day when two chillers are active and the valve signal reduces to ~0.5, it reduces for both parallel valves simultaneously.

<img width="597" alt="Screen Shot 2023-05-31 at 3 48 33 PM" src="https://github.com/lbl-srg/modelica-buildings/assets/19480999/fbabf530-be06-4546-9d27-0dd90d47c7b0">

So, I think this assumption is ok.

In the end, I think it's best not to make any modifications to how the multiple valves are implemented.

dhblum commented:
@mwetter The unit tests are failing due to stalled builds and timeouts for openmodelica [here](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/603306802) and [here](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/603306775).  Have you seen this before?

mwetter commented:
I deleted the cache and restarted the failed job. If the individual simulation times out we should try to increase the timeout in the conf.yml file.

dhblum commented:
Thanks @mwetter.  Another thought I had is if some added tests in the Experimental package are now making the collection of Buildings.{Applications,Experimental} take too long in ``TEST_ARG="make test-openmodelica PACKAGE=\"Buildings.{Applications,Experimental}\""``, and Applications and Experimental should be split into different test args?

mwetter commented:
@dhblum : Feel free to split these into two packages, and use the same split for all 3 tools that we test.

dhblum commented:
@mwetter A couple models are failing Dymola and Optimica unit tests which use Spawn [here](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/603548221) and [here](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/603548223).  But I can't repeat those errors locally, the models run just fine with Dymola and Optimica.  Any ideas?

JayHuLBL commented:
@dhblum Just echo you. I just run locally and also cannot reproduce the errors.

dhblum commented:
Thanks @JayHuLBL!

hcasperfu commented:
@dhblum -
I ran the unit tests locally and the two same models from CI failed. Then I tried to run them directly in Dymola, which gave me the message that Spawn wasn't installed (which was odd because I thought I did). I installed Spawn, and reran the test scripts, and now my tests pass.

dhblum commented:
Thanks for looking into this @hcasperfu.  Did you run the unit tests initially (when they failed) with a fresh clone/copy of the Buildings Library?

hcasperfu commented:
@dhblum  - 
I made a fresh clone of Buildings and the tests failed the same way as I checked out this branch to my existing local repo.
However I notice the errors aren't the same as [those seen in the CI](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/604189585). In my `unitTests-dymola.log` there is no `IOError`, but a few of the following:
```
*** Error: When processing /tmp/tmp-Buildings-0-rhm5iw1x/Buildings/Buildings.Experimental.DHC.Loads.BaseClasses.Examples.CouplingSpawnZ1.mat generated by Experimental/DHC/Loads/BaseClasses/Examples/CouplingSpawnZ1.mos, caught division by zero.
   len(val)  = 1
   tMax-tMin = 0.0
```
So I'm starting to suspect the test failure generated on my machine is a different one.

dhblum commented:
I think I found the issue with the sudden failing of Spawn models in travis unit tests!  See https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/69e90b4f5d6afafb855e7d9135099b296757e5cc, needed in response to https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/8468215a9828af0b665b81ab95e7c24f3cace0f7.

mwetter commented:
Good catch!

On Wed, Jul 12, 2023 at 1:26 PM David Blum ***@***.***> wrote:

> I think I found the issue with the sudden failing of Spawn models in
> travis unit tests! See 69e90b4
> <https://github.com/lbl-srg/modelica-buildings/commit/69e90b4f5d6afafb855e7d9135099b296757e5cc>,
> needed in response to 8468215
> <https://github.com/lbl-srg/modelica-buildings/commit/8468215a9828af0b665b81ab95e7c24f3cace0f7>
> .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lbl-srg/modelica-buildings/pull/3263#issuecomment-1633164215>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEPODRUHGFWCQNHWVK7BLXP4B6LANCNFSM6AAAAAAVHFL3UA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dhblum commented:
Some documentation and thoughts on why unit tests are failing at https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/9a94dce0a903fe2395d1d275dc7c59720114a81f.  @mwetter I appreciate any thoughts you have when you get a chance:

- [Job .35](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/606207008) is failing because of a message that the command ``returned non-zero exit status 137``.  This makes me think the process is running out of memory.  Running ``$ python ../bin/runUnitTests.py --batch --single-package "Buildings.Experimental.DHC.Examples.Combined" --tool openmodelica --skip-verification`` locally passes ok, but uses > 7 GB of memory + up to 1 GB of swap.  Forcing the tests to use only 1 processor (adding ``-n 1``) uses only 4.5 GB memory, though takes longer to complete both tests.  I'm testing whether forcing use of 1 processor helps in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/3e8cf26735090800c6e98ec55a96aa4a0aac5247, but this is intrusive because it requires use of a new argument in the TEST_ARG etc. to specify number of processors to use.  Note also I moved the test to the top of the list just so I don't have to wait too long to see if it works.  But in any case it seems its in a queue due to other PRs/commits in the library.

- [Job .33](https://app.travis-ci.com/github/lbl-srg/modelica-buildings/jobs/606207006) is failing due to ``RuntimeError: Wrong invocation, generated 0 unit tests.``.  The job only has 1 example model (split from previous implementation in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/9a94dce0a903fe2395d1d275dc7c59720114a81f to avoid timeout), but that model is on the exclude list for translation for Optimica so I wonder if it's trying to run a test when actually none are to be run, and that's an error not handled in buildingspy?  The output does print ``Using 0 of 2 processors to run unit tests for optimica.``, which is maybe a no-no right now.

dhblum commented:
@mwetter In summary, tests are now passing.  But this includes a new modifier added in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/750abf674dc750fae06c0d658ed06f442255c743 to enable tests to run one processor at a time in travis to not cause memory timeout.  Your thoughts appreciated, maybe a better way to do it.  Otherwise, this is ready to go I think.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Controls/BaseClasses/PartialController.mo:
One reason might be that it defines the controller interface, which could be replaced with different logic in a new or variation controller down the line or in future studies.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Controls/BaseClasses/StagingPump.mo:
Addressed in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/a099f9569fa1abeac5eba6b04d6e47625004e47e.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Controls/BaseClasses/CoolingTowerLoop.mo:
Addressed in https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/a099f9569fa1abeac5eba6b04d6e47625004e47e.

dhblum commented:
Unit tests are failing for:

```
*** Error: Translation of Buildings.Obsolete.Examples.VAVReheat.Validation.Guideline36SteadyState failed: 'TimeoutExpired: Command '['omc', 'Buildings.Obsolete.Examples.VAVReheat.Validation.Guideline36SteadyState_translate.mos']' timed out after 300 seconds'. Directory is '/tmp/tmp-Buildings-0-511guoxa'.
Execution time = 2329.967 s
make: *** [Makefile:167: test-openmodelica] Error 1
The command "(cd Buildings/Resources/Scripts/travis && travis_wait 90 $TEST_ARG)" exited with 2.
```

@mwetter  This is unrelated to any changes in the PR as far as I can tell.

mwetter commented:
@dhblum Can you improve the time out in conf.yml to 600?

dhblum commented:
Thanks @mwetter.  The time_out parameter in conf.yml is new to me, but I think I did it right here: https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/e8ba5cc129cfb64e347be9569c11f7bf662d1526.

dhblum commented:
Huzzah, unit tests pass.

dhblum commented on Buildings/Experimental/DHC/Plants/Combined/Controls/BaseClasses/CoolingTowerLoop.mo:
```diff
@@ -0,0 +1,529 @@
+    "Stage pumps"
+    annotation (Placement(transformation(extent={{100,50},{120,70}})));
+  Buildings.Controls.OBC.CDL.Continuous.GreaterThreshold cmpFlo(t=3, h=3e-3)
+    "Flow criterion to enable lead pump"
```
@mwetter I addressed this and other instances as we discussed in commits https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/1fc7b11c1974f398a562956c72cd201faf6abfe1 through https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/41a7c3177183734a75bb3bb16d853fe8e50170f7.  Note that the threshold and hysteresis parameters of the greater/less than CDL block can't be dependent on iteration variables of the initialization problem, so [here](https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/b6ca2abfe5a55d72691ec556bfdf626edd1fbd66#diff-6b12c02182a8e2551902e7fbf229c7c7b0bb5a9d3ad583a5bdeb16437f6e8758R216) I needed to use a hard-coded cp and [here](https://github.com/lbl-srg/modelica-buildings/pull/3263/commits/b6ca2abfe5a55d72691ec556bfdf626edd1fbd66#diff-bfb29305f9a33a5b446e9b82a281a718287d8d10715bcb5ddc6c071fb0030221R235) I needed to introduce an approximate value of the nominal cooling capacity of the heat recovery chillers in cascade cooling mode which is directly specified (basically as datChiHea.QEva_flow_nominal), instead of the exact value that is found through iteration at the initialization.

dhblum commented:
@mwetter FYI despite my latest commits to address the hysteresis, this PR is not ready to merge anymore.  I'm doing some testing to simulate the plant for a fully year and it's not finishing.  Need to look into this again.

dhblum commented:
@mwetter This morning I tested again the example model which was failing the full simulation before, and after merging latest master,  it is now working and simulated the full year ok with dassl and tol 10e-6.  I'm not sure what happened before, maybe it was something with my setup at the time.  I think I addressed all the hysteresis issues and your comments, so except for confirming unit test pass, let me know if you think there's anything else to be done.  I don't have anything in mind.  @AntoineGautier might have a review at it as he is back.
